### PR TITLE
Bug 2000073: Show "project not found" error if a namespace selected via an URL doesn't exist

### PR DESCRIPTION
**Pull request by jerolimov**
**Fixes**: 
Bug https://bugzilla.redhat.com/show_bug.cgi?id=2000073
Similar reported as https://issues.redhat.com/browse/ODC-4474

**Analysis / Root cause**: 
The selected namespaced is saved in the redux store and there is no detection if another user (or the same in another tab) removes this namespace. The user can navigate between multiple screen without notice that his namespace doesn't exist.

(Btw: Our / the k8s API returns an empty array if the user asks for a non-existing namespace "pod list".)

If the latest or preferred namespace wasn't found, and if the user opens `/`, `/add`, `/topology`, etc., and the preferred or latest namespace is removed, the user will automatically switch back to pages that require the user to select a namespace.

But when the user opens a specific namespace via an URL, like `/add/ns/my-namespace`, `/topology/ns/my-namespace`, etc., this check isn't implemented.

But IMHO this is correct and should not be changed. When the user opens a bookmark, or a link from another associate, called "MY-TEST-NAMESPACE topology", we should not switch the namespace. In the worst case, the user didn't notice that the new namespace is MY-PRODUCTION-NAMESPACE"...

I assumed that this is an edge-case. I decided to NOT wait until the complete project list is fetched and also show the normal page if there is any error while fetching this data. So I hope this will also help in some edge-cases where the user could not see their own project but can create a resource? I'm unsure if this could happen. :smile: 

**Solution Description**: 
This change shows instead a "404 Project/Namespace not found" page if the user opens a namespace that doesn't exist.

## Similar to the cases where a resource is not found:

image

## New screen when the namespace in the URL doesn't exist:

image

This check is added to these shared components: NamespacedPage, ResourceListPage and ResourceDetailPage.

Because these three pages already fetch the complete project/namespace list for the `NamespaceDropdown`. So instead of another API call the "exist check" reuse the same data.

**Screenshots / Gifs for design review**: 

## Screen recording which shows different case before:

https://user-images.githubusercontent.com/139310/155741798-a34ed51e-1dd5-4707-a8fe-5076cec75c36.mp4

## Screen recording which shows different case after:

https://user-images.githubusercontent.com/139310/155741821-2dfd78c2-7997-40ae-907e-cae05393dc7a.mp4

## Tested some more screens:

You can see that there are some flickers when switching between pages in the dev console, that happens because each page reloads the project list again and again. I don't want to change this behavior as part of this ticket. I think this could be improved, but as said, this PR here should address more the edge case that the user opens a link with a non-existing project...

https://user-images.githubusercontent.com/139310/155741889-569da751-a895-40b1-a834-e6feaa819c39.mp4

**Unit test coverage report**: 
Added complete new tests

**Test setup:**
1. Open a URL that specifies a namespace, try existing and non-existing projects.
2. Try some edge cases.

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Comment by jerolimov** [comment]
Set this on hold to get feedback here also from Sam and Serena.

Can you please take a look if this solution makes sense to you?

Tl;dr-summary: Checking the namespace and showing an error if the namespace doesn't exist. This new check happens only if the user opens a URL **with a namespace.**

If the user opens just the console, or /add as URL the browser navigates to the latest (or preferred) namespace. If the latest or preferred namespace was removed in the mean time, the user was redirected to "all namespaces" as before.

/cc @spadgett @serenamarie125 
/hold

**Review by spadgett — CHANGES_REQUESTED** [review]
It looks to me like this will break deep-linking to resources in namespaces that aren't in your project list. We can't assume that just because the namespace isn't in your project list that you can't see resources in that namespace. (We've had this bug a few times in the past and rejected fixing it for this reason.)

/hold

**Comment by jerolimov** [comment]
> It looks to me like this will break deep-linking to resources in namespaces that aren't in your project list. We can't assume that just because the namespace isn't in your project list that you can't see resources in that namespace. (We've had this bug a few times in the past and rejected fixing it for this reason.)

OK, that's the reason I want your review. I wasn't aware that this was possible and expected. But totally makes sense to me. Thanks for sharing that.

So, @spadgett @serenamarie125, do you think we should check for the current namespace on this pages

* "Create forms" as shown in https://bugzilla.redhat.com/show_bug.cgi?id=2000073
    * `.../~new` and `..../~new/...` URLs (includes also pipeline builder)
* And on the add page for https://issues.redhat.com/browse/ODC-4474
    * `/add/ns/...` (Add page)
    * `/catalog/ns/...` (Add > Catalog pages)
    * `/samples/ns/...` (Add > Samples)
    * `/import/ns/...` (Add > Import from git)
    * `/deploy-image/ns/...` (Add > Container images)
    * `/upload-jar/ns/...` (Add > Upload JAR flow)
    * `/channel/ns/...` (Add > kn Channel)
    * `/broker/ns/...` (Add > kn Broker)

And then show such a "404 Project not found" page  -OR- should we just show a warning box at the top of the form? :thinking: 

I would still add this check to the individual called page. I just added the URLs to that the pages are clear and just to confirm that this 2nd approach goes into right direction.

Wdyt?
